### PR TITLE
Base Gemma diffusion implementation

### PR DESCRIPTION
**Pull request by aleroot**
## Proposed changes

Basic Gemma diffusion model implementation.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: Libraries/MLXLMCommon/Evaluate.swift [hunk 1]
```diff
@@ -707,6 +767,314 @@ public struct TokenIterator: TokenIteratorProtocol {
     }
 }
 
+/// Generator for block-diffusion language models.
+///
+/// The model first encodes the prompt into a KV cache, then repeatedly denoises
+/// a full canvas. The iterator exposes the finalized canvas tokens one by one
+/// so existing detokenization, EOS handling, tool parsing, and timing code can
+/// remain shared with autoregressive iterators.
+public struct BlockDiffusionTokenIterator: TokenIteratorProtocol {
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/Evaluate.swift]: This looks great, but i wonder if this should be in its own file?  Evaluate.swift is getting unwieldy.

File: Libraries/MLXLMCommon/Evaluate.swift [hunk 2]
```diff
@@ -778,6 +1146,15 @@ public struct SpeculativeTokenIterator: TokenIteratorProtocol {
...
         numDraftTokens: Int
     ) throws {
+        guard !(mainModel is any BlockDiffusionLanguageModel) else {
+            throw GenerateError.unsupportedSpeculativeDecoding(
+                String(describing: type(of: mainModel)))
+        }
+        guard !(draftModel is any BlockDiffusionLanguageModel) else {
+            throw GenerateError.unsupportedSpeculativeDecoding(
+                String(describing: type(of: draftModel)))
+        }
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/Evaluate.swift]: I think this is fine for now but we might want some kind of capability property down the road.  maybe long these lines:

https://developer.apple.com/documentation/foundationmodels/languagemodelcapabilities/capability/vision

then we can test directly vs using `is` / `as`

File: Libraries/MLXLMCommon/KVCache.swift [hunk 3]
```diff
@@ -669,7 +669,7 @@ public class RotatingKVCache: BaseKVCache, CustomDebugStringConvertible {
                     values[.ellipsis, ..<offset, 0...],
                 ]
             } else {
-                return [keys, values]
+                return [temporalOrder(keys), temporalOrder(values)]
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/KVCache.swift]: I think this is desired for this diffusion model, but what about other users of RotatingKVCache?  this will be a semantic change for them.  Also this interacts with serialization.

Maybe it is fine -- this is used in one path for updates, but i think it needs closer examination.

File: Libraries/MLXLMCommon/Evaluate.swift [hunk 4]
```diff
@@ -707,6 +767,314 @@ public struct TokenIterator: TokenIteratorProtocol {
...
+    private mutating func processLogits(_ logits: MLXArray, curStep: Int) -> MLXArray {
+        let temperature = MLXArray(temperature(curStep: curStep))
+        guard processor != nil else {
+            return logits / temperature
+        }
+
+        let length = logits.dim(1)
+        let processed = (0 ..< length).map { index in
+            var tokenLogits = logits[0..., index, 0...]
+            tokenLogits = processor?.process(logits: tokenLogits) ?? tokenLogits
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/Evaluate.swift]: It looks like the mlx-vlm side doesn't even support logit processors:

```python
                schedule_temperature = _diffusion_linear_temperature(
                    cur_step,
                    max_denoising_steps,
                    temperature_config,
                )
                if schedule_temperature is not None:
                    processed_logits = processed_logits / schedule_temperature

                argmax_canvas = mx.argmax(processed_logits, axis=-1).astype(
                    input_ids.dtype
                )
                if cur_step == 1 and not diffusion_show_unmasking:
                    break

                denoiser_canvas = (
                    argmax_canvas
                    if temperature <= 0
                    else _diffusion_sample_canvas(
                        processed_logits,
                        input_ids.dtype,
                        temperature,
                    )
                )
```

I am not sure this is safe to do -- the logit processors can be stateful (e.g. repetition penalties) and applying them to the entire canvas would (I think) leak state into the processor even if we don't end up using the token.

Check my math on that, but if it is correct do we:

- it is fine
- need a notion of a stateless logit processor that could be applied?
- be able to back out logits that do not make it?
- apply that processor after we accept the logits?

I am not sure I fully wrap my head around the diffusion path so I might be off base here.

Comment (aleroot) [reply Libraries/MLXLMCommon/Evaluate.swift]: This is an important point. For the moment I will remove it, but in the future there should probably be an explicit new API:

- StatelessLogitProcessor: safe for pure masks/biases that do not depend on generated history.
- DiffusionLogitProcessor: receives canvas position, reveal mask, accepted mask/history, denoising step, etc.

Apply penalties only after token acceptance: safe for state, but too late to affect selection, so mostly useless for logits processing.

File: Libraries/MLXLMCommon/KVCache.swift (hunk 3 above)
Comment (aleroot) [reply Libraries/MLXLMCommon/KVCache.swift]: Good point. I changed this so `RotatingKVCache.state` keeps the existing ring-buffer storage order.

File: Libraries/MLXLMCommon/BlockDiffusionTokenIterator.swift [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+                currentCanvas = MLX.where(
+                    draftRevealMask | acceptedMask,
+                    acceptedCanvas,
+                    makeInitialCanvas(length: length))
+                draftRevealMask = draftRevealMask | acceptedMask
+                draftCanvas = acceptedCanvas
+
+                if draftRevealMask.all().item(Bool.self) {
+                    argmaxCanvas = draftCanvas
+                    break
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/BlockDiffusionTokenIterator.swift]: I wonder if this break is meant to break out of the loop?  I think this is the matching python code:

```python
                if diffusion_sampler == "confidence-threshold" and bool(
                    mx.all(draft_reveal_mask).item()
                ):
                    accepted_canvas = draft_canvas
                    break
```

I think since this swift code is in a switch it only breaks the switch itself, not an exit from the loop.

File: Libraries/MLXVLM/Models/DiffusionGemma.swift [hunk 6]
```diff
@@ -0,0 +1,701 @@
...
+        DiffusionGemmaConfiguration(
+            textConfig: textConfig,
+            generationConfig: generationConfig,
+            canvasLength: canvasLength,
+            tieWordEmbeddings: tieWordEmbeddings,
+            modelType: modelType)
+    }
+}
+
+public final class DiffusionGemma: Module, VLMModel, BlockDiffusionLanguageModel,
```
Comment (davidkoski) [inline Libraries/MLXVLM/Models/DiffusionGemma.swift:88]: Does this support tools?  It doesn't define a `toolCallFormat`

File: Libraries/MLXLMCommon/Models/DiffusionGemma.swift [hunk 7]
```diff
@@ -0,0 +1,1167 @@
...
+        selfConditioningEmbeddings: MLXArray? = nil
+    ) -> MLXArray {
+        var h = embeddings(canvasTokens)
+
+        let signal: MLXArray
+        if let selfConditioningEmbeddings {
+            signal = selfConditioningEmbeddings.asType(h.dtype)
+        } else if let selfConditioningLogits {
+            let probabilities = softmax(
+                selfConditioningLogits.asType(h.dtype), axis: -1, precise: true)
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/Models/DiffusionGemma.swift]: Spotted by claude in comparison with mlx-vlm.  The corresponding python code is:

```python
    def _embed_canvas(
        self,
        canvas_ids,
        self_conditioning_logits=None,
        self_conditioning_embeddings=None,
    ):
        inputs_embeds = self.embed_tokens(canvas_ids) * self.embed_scale
..
        else:
            probs = mx.softmax(self_conditioning_logits, axis=-1, precise=True)
```

notice the lack of type conversion.  The is another python function that does match this:

```python
    def diffusion_self_conditioning(
        self,
        processed_logits: mx.array,
        embedding_weight: Optional[mx.array],
    ) -> mx.array:
        if self.prefers_logits_self_conditioning:
            return processed_logits
        # Match the HF generation path: self-conditioning logits are stored in
        # the embedding dtype before the next decoder softmax.
        probs = mx.softmax(
            processed_logits.astype(embedding_weight.dtype),
            axis=-1,
            precise=True,
        )
```

but I think that matches `selfConditioningEmbeddings()` in this file.

File: Libraries/MLXLMCommon/Models/DiffusionGemma.swift [hunk 8]
```diff
@@ -0,0 +1,1167 @@
...
+    }
+
+    let state =
+        if let rotatingCache = cache as? RotatingKVCache {
+            rotatingCache.temporalState
+        } else {
+            cache.state
+        }
+
+    guard state.count == 2 else {
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/Models/DiffusionGemma.swift]: Can this model be configured with a QuantizedKVCache?  That can return state with more than 2 elements.

The state is really meant for doing `eval`, I wonder if we need API that exposes this properly?  The ordering of the values in state is not specified.

File: Libraries/MLXLMCommon/Evaluate.swift [hunk 9]
```diff
@@ -1057,6 +1138,15 @@ public struct SpeculativeTokenIterator: TokenIteratorProtocol {
         numDraftTokens: Int,
         components: GenerationComponents = .init()
     ) throws {
+        guard !mainModel.capabilities.contains(.blockDiffusion) else {
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/Evaluate.swift:1141]: I think MTPSpeculativeTokenIterator may need the same checks.
